# Section 12.3, Exercises 12.15-12.17

Nathan Schill

In [13]:
import numpy as np
from scipy import optimize

Df = optimize.approx_fprime
newton = optimize.newton

## 12.15 - Exact Gradient Descent for Quadratic Functions

In [3]:
def quad_grad_descent(Q, b, x0, eps=1e-6):
    # Max number of iterations in case of failure to converge
    max_iter = 1000

    # Compute xk using the exact gradient descent algorithm
    xk = x0
    for _ in range(max_iter):
        # Get Df(xk)^T
        DfxkT = Q@xk - b

        # Return xk if ||DfxkT|| < eps
        if np.linalg.norm(DfxkT) < eps:
            return xk

        # Compute next iteration
        alpha = (DfxkT.T@DfxkT) / (DfxkT.T@Q@DfxkT)
        xk = xk - alpha * DfxkT

    return None      

Q = np.array([[2, -2],
              [-2, 4]])
b = np.array([-3, 4])
x0 = np.array([2,0])

quad_grad_descent(Q, b, x0)

array([-0.99999953,  0.50000031])

## 12.16 - Exact Gradient Descent (general)

In [50]:
def line_search(f, xk, DfxkT):
    '''Return a critical point alpha which we hope is a minimizer.'''
    # Initial alpha guess: Use this because Example 12.3.2
    # found alpha=0.1 to be a minimizer for its problem
    a0 = 0.1
    
    # Want to minimize f(xk - a*DfxkT) wrt a
    g = lambda a: f(xk - a*DfxkT)

    # Create function that approximates derivative of g at the input
    tol = np.sqrt(np.finfo(float).eps)
    Dg = lambda x: Df(x, g, tol)

    # Return the minimizer alpha
    return newton(Dg, a0, maxiter=100, disp=False)


def exact_grad_descent(f, x0, eps=1e-6):
    # Max number of iterations in case of failure to converge
    max_iter = 10000

    # tol to use for finite difference quotient approximation of Df(xk)
    tol = np.sqrt(np.finfo(float).eps)

    # Compute xk using the exact gradient descent algorithm
    xk = x0
    for _ in range(max_iter):
        # Get Df(xk)^T
        DfxkT = Df(xk, f, tol).T

        # Return xk if ||DfxkT|| < eps
        if np.linalg.norm(DfxkT) < eps:
            return xk

        # Compute next iteration
        alpha = line_search(f, xk, DfxkT)
        xk = xk - alpha * DfxkT
        
    return 'Did not converge'


def f(x):
    Q = np.array([[2, -2],
              [-2, 4]])
    b = np.array([-3, 4])
    return 1/2 * x@Q@x - b@x + 5

x0 = np.array([2,0])
f(x0)

exact_grad_descent(f, x0)

array([-0.99999898,  0.50000063])

## 12.17 - Exact G.D. on Rosenbrock function

In [48]:
# Modify function for this problem to return when estimate is within 1e-5 of the true minimizer (1,1)
# since the function from the previous problem checks the norm of the derivative...
def exact_grad_descent(f, x0, eps=1e-6):
    true_minimizer = np.array([1,1])

    # Max number of iterations in case of failure to converge
    max_iter = 10000

    # tol to use for finite difference quotient approximation of Df(xk)
    tol = np.sqrt(np.finfo(float).eps)

    # Compute xk using the exact gradient descent algorithm
    xk = x0
    for i in range(max_iter):
        # Get Df(xk)^T
        DfxkT = Df(xk, f, tol).T

        # Return xk if ||DfxkT|| < eps
        if np.linalg.norm(xk - true_minimizer) < 1e-5 or np.linalg.norm(DfxkT) < eps:
            print('Num iters:', i)
            return xk

        # Compute next iteration
        alpha = line_search(f, xk, DfxkT)
        xk = xk - alpha * DfxkT
        
    return 'Did not converge'

# Try running it here!
def f(t):
    x, y = t
    return 100*(y - x**2)**2 + (1 - x)**2

exact_grad_descent(f, (-2,2))

Num iters: 8230


array([0.9999999 , 1.00000015])

#### Does it converge? If not, explain why not. If it does, how many iterations does it take to get within 1e-5 of the true minimizer?

It converged in 8230 iterations